src/main/java/io/dockstore/provision/GSPlugin.java

```diff
@@ -0,0 +1,271 @@
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+package io.dockstore.provision;
+
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.PrintStream;
+import java.lang.reflect.Type;
+import java.nio.ByteBuffer;
+import java.nio.channels.Channels;
+import java.nio.channels.WritableByteChannel;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+import com.google.cloud.ReadChannel;
+import com.google.cloud.WriteChannel;
+import com.google.cloud.storage.Blob;
+import com.google.cloud.storage.BlobId;
+import com.google.cloud.storage.BlobInfo;
+import com.google.cloud.storage.Storage;
+import com.google.cloud.storage.StorageException;
+import com.google.cloud.storage.StorageOptions;
+import com.google.common.collect.Lists;
+import com.google.gson.Gson;
+import com.google.gson.reflect.TypeToken;
+import org.apache.commons.lang3.StringUtils;
+import ro.fortsoft.pf4j.Extension;
+import ro.fortsoft.pf4j.Plugin;
+import ro.fortsoft.pf4j.PluginWrapper;
+import ro.fortsoft.pf4j.RuntimeMode;
+
+// Imports the Google Cloud client library
+
+/**
+ * @author wshands
+ */
+public class GSPlugin extends Plugin {
+
+    public GSPlugin(PluginWrapper wrapper) {
+        super(wrapper);
+    }
+
+    @Override
+    public void start() {
+        // for testing the development mode
+        if (RuntimeMode.DEVELOPMENT.equals(wrapper.getRuntimeMode())) {
+            System.out.println(StringUtils.upperCase("GSPlugin development mode"));
+        }
+    }
+
+    @Override
+    public void stop() {
+        System.out.println("GSPlugin.stop()");
+    }
+
+    @Extension
+    public static class GSProvision implements ProvisionInterface {
+
+        private static final int MAX_FILE_SIZE_FOR_SINGLE_WRITE = 1_000_000;
+        private static final int DOWNLOAD_BUFFER_SIZE = 64 * 1024;
+        private static final int UPLOAD_BUFFER_SIZE = 1024;
+        private Map<String, String> config;
+
+        public void setConfiguration(Map<String, String> map) {
+            this.config = map;
+        }
+
+        private Storage getGoogleGSClient() {
+            return StorageOptions.getDefaultInstance().getService();
+        }
+
+        public Set<String> schemesHandled() {
+            return new HashSet<>(Lists.newArrayList("gs"));
+        }
+
+        public boolean downloadFrom(String sourcePath, Path destination) {
+            Storage gsClient = getGoogleGSClient();
+
+            String trimmedPath = sourcePath.replace("gs://", "");
```

Comment:
I would pull this out into a function and write a unit test for it.

Reply:
Refactored to separate method and test added